test: Add remaining scenarios of 0 waste, in wallet waste_test

rajarshimaitra opened the pull request:
As per the [review club](https://bitcoincore.reviews/22009) discussion on #22009 , it was observed that there were other two fee scenarios in which selection waste could be zero.

These are:
 - (LTF - Fee) == Change Cost
 - (LTF - Fee) == Excess

Even though these are obvious by the definition of waste metric, adding tests for them can be helpful in explaining its behavior
to new readers of the code base, along with pinning the behavior for future.

This PR adds those two cases to waste calculation unit test.

Also let me know if I am missing more scenarios.

rajarshimaitra committed:
test: Add remaining scenarios of 0 waste

There are two more cases where waste can be 0, when:
 - (Fee - LTF) == -Change Cost
 - (Fee - LTF) == -Excess

Adding these two conditions explicitly in the unit test will help
pin the behavior, also demonstrate waste calculation scenarios to new
readers.

theStack commented:
Concept ACK

Xekyo commented:
Concept ACK

> These are:
> 
>     * (Fee - LTF) == Change Cost
> 
>     * (Fee - LTF) == Excess

Strictly speaking it should be: 
   * (Fee - LTF) + Change Cost = 0  
      ⇒ (Fee - LTF) == -1×Change Cost  

   * (Fee - LTF) + Excess = 0
     ⇒ (Fee - LTF) == -1×Excess

achow101 commented:
ACK c976ee0a0ca00fe8a7de5c806560da35c2f48729

Xekyo commented on src/wallet/test/coinselector_tests.cpp:
```diff
@@ -724,12 +724,25 @@ BOOST_AUTO_TEST_CASE(waste_test)
     BOOST_CHECK_EQUAL(0, GetSelectionWaste(selection, 0, exact_target));
+    selection.clear();
 
+    // 0 Waste when (fee - long term fee) == cost of change, and no excess
```
```suggestion
    // 0 Waste when (long term fee - fee) == cost of change, and no excess
```
As it is currently written the two aren't equal but negatives of each other.

Xekyo commented on src/wallet/test/coinselector_tests.cpp:
```diff
@@ -724,12 +724,25 @@ BOOST_AUTO_TEST_CASE(waste_test)
+    BOOST_CHECK_EQUAL(0, GetSelectionWaste(selection, new_change_cost, target));
+    selection.clear();
+
+    // 0 Waste when (fee - long term fee) == excess, no change cost
```
```suggestion
    // 0 Waste when (long term fee - fee) == excess, no change cost
```
Same here.

murchandamus reviewed (commented):
Left a nit on the comment explaining the test

rajarshimaitra commented:
Thanks @Xekyo for the review and agreed. It should be reversed. 

Updated everywhere..

jonatack commented on src/wallet/test/coinselector_tests.cpp:
```diff
@@ -724,12 +724,25 @@ BOOST_AUTO_TEST_CASE(waste_test)
+    const CAmount new_target = in_amt - fee * 2 - fee_diff * 2;
+    add_coin(1 * COIN, 1, selection, fee, fee + fee_diff);
+    add_coin(2 * COIN, 2, selection, fee, fee + fee_diff);
+    BOOST_CHECK_EQUAL(0, GetSelectionWaste(selection, 0, new_target));
```
minor suggestions: `s/0 Waste/No waste/`, braced initialization for type safety, named args
```diff
@@ -724,25 +724,25 @@ BOOST_AUTO_TEST_CASE(waste_test)
     BOOST_CHECK_LT(waste_nochange2, waste_nochange1);
     selection.clear();
 
-    // 0 Waste when fee == long term fee, no change, and no excess
+    // No waste when fee == long term fee, no change, and no excess
     add_coin(1 * COIN, 1, selection, fee, fee);
     add_coin(2 * COIN, 2, selection, fee, fee);
-    const CAmount exact_target = in_amt - 2 * fee;
-    BOOST_CHECK_EQUAL(0, GetSelectionWaste(selection, 0, exact_target));
+    const CAmount exact_target{in_amt - 2 * fee};
+    BOOST_CHECK_EQUAL(0, GetSelectionWaste(selection, /* change_cost */ 0, exact_target));
     selection.clear();
 
-    // 0 Waste when (long term fee - fee) == cost of change, and no excess
-    const CAmount new_change_cost = fee_diff * 2;
+    // No waste when (long term fee - fee) == cost of change, and no excess
+    const CAmount new_change_cost{fee_diff * 2};
     add_coin(1 * COIN, 1, selection, fee, fee + fee_diff);
     add_coin(2 * COIN, 2, selection, fee, fee + fee_diff);
     BOOST_CHECK_EQUAL(0, GetSelectionWaste(selection, new_change_cost, target));
     selection.clear();
 
-    // 0 Waste when (long term fee - fee) == excess, no change cost
-    const CAmount new_target = in_amt - fee * 2 - fee_diff * 2;
+    // No waste when (long term fee - fee) == excess, no change cost
+    const CAmount new_target{in_amt - fee * 2 - fee_diff * 2};
     add_coin(1 * COIN, 1, selection, fee, fee + fee_diff);
     add_coin(2 * COIN, 2, selection, fee, fee + fee_diff);
-    BOOST_CHECK_EQUAL(0, GetSelectionWaste(selection, 0, new_target));
+    BOOST_CHECK_EQUAL(0, GetSelectionWaste(selection, /* change_cost */ 0, new_target));
```

jonatack reviewed (commented):
ACK 0ce33016eb905f5ac3709ed95ae02db0f97885a2

Perhaps update the commit message and pull description with @Xekyo's calculations in https://github.com/bitcoin/bitcoin/pull/22938#issuecomment-916948224. Edit: never mind, it seems fine now IIUC.

Xekyo commented:
ACK 0ce33016eb905f5ac3709ed95ae02db0f97885a2

rajarshimaitra commented:
Thanks @jonatack for the review. Those were good suggestions.

Updated the comment description.
Updated commit message.
Updated with braced initialization.

Xekyo commented:
Re-ack https://github.com/bitcoin/bitcoin/commit/a2a9c23e261ad88c0133b407c6e1b4c54f90a860

jonatack commented:
Code review re-ACK a2a9c23e261ad88c0133b407c6e1b4c54f90a860 per `git diff 0ce3301 a2a9c23`

jonatack commented on src/wallet/test/coinselector_tests.cpp:
```diff
@@ -724,12 +724,25 @@ BOOST_AUTO_TEST_CASE(waste_test)
     add_coin(2 * COIN, 2, selection, fee, fee);
-    const CAmount exact_target = in_amt - 2 * fee;
-    BOOST_CHECK_EQUAL(0, GetSelectionWaste(selection, 0, exact_target));
+    const CAmount exact_target{in_amt - 2 * fee};
```
```suggestion
    const CAmount exact_target{in_amt - fee * 2};
```

jonatack commented on src/wallet/test/coinselector_tests.cpp:
```diff
@@ -724,12 +724,25 @@ BOOST_AUTO_TEST_CASE(waste_test)
+    BOOST_CHECK_EQUAL(0, GetSelectionWaste(selection, new_change_cost, target));
+    selection.clear();
+
+    // 0 Waste when (fee - long_term_fee) == (-excess), no change cost
```
```suggestion
    // No Waste when (fee - long_term_fee) == (-excess), no change cost
```

jonatack reviewed (commented):
Code review re-ACK a2a9c23e261ad88c0133b407c6e1b4c54f90a860 per `git diff 0ce3301 a2a9c23`


Sub-atomic consistency nits, but definitely ignore unless you have to rebase or something (and maybe even then).

rajarshimaitra commented:
@jonatack Sorry for my silly overlooks.. Updated..

jonatack commented:
Tested re-ACK efcaefc7b5ffe0495e7c809032342ee5ca4841be

achow101 commented:
ACK efcaefc7b5ffe0495e7c809032342ee5ca4841be

meshcollider reviewed (commented):
ACK efcaefc7b5ffe0495e7c809032342ee5ca4841be
